evap/results/tools.py

```diff
@@ -41,28 +41,42 @@ def __init__(self, visible_by_contribution: Iterable[UserProfile], visible_by_de
         self.visible_by_delegation_count = visible_by_delegation_count
 
 
+def create_rating_result(question, answer_counters, additional_text_result=None):
+    if answer_counters is None:
+        return RatingResult(question, answer_counters, additional_text_result)
+    if any(counter.count != 0 for counter in answer_counters):
+        return AnsweredRatingResult(question, answer_counters, additional_text_result)
+    return PublishedRatingResult(question, answer_counters, additional_text_result)
+
+
 class RatingResult:
+    is_published = False
+    has_answers = False
```

Comment:
This might be a funky idea, but maybe the following is possible:

```python
class RatingResult:
    @property
    def is_published(self) -> TypeGuard["PublishedRatingResult"]:
        return isinstance(self, PublishedRatingResult)
```
I would consider this a faithful use of `TypeGuard` and it would allow the results to be used exactly the same in caller code, but with the nice underlying structure of this PR (I like in particular that the methods are not even available on `RatingResult` until you have checked the type!). What do you think?

Reply:
sadly, the auther of the PEP decided not support this. Only the first explicitly passed parameter can be guarded :cry: 
https://github.com/microsoft/pyright/discussions/3125#discussioncomment-2266438
https://github.com/python/mypy/issues/16618#issuecomment-1840650889

So, as far as I can see, we either have to use `isinstance` directly or via `is_published(...)`, but not via an instance method, especially not with a property

Reply:
I'd be fine with making this a freestanding or static method if we leave a comment to the docs (I'd link https://peps.python.org/pep-0647/#narrowing-of-implicit-self-and-cls-parameters).

Reply:
the problem is, that the templates need to access those too. With that in mind, there are no use cases for a typeguarded `has_answers` function without #2035. `is_published` is only used inside templates (once).

Reply:
That's too bad, I really thought I was onto something here :D

For templates, we could also register `has_answers` etc. as template filters (we can still define them next to the classes and only register them in `evaluation_filters.py`).

We need a type guard in some way though (I am still convinced that splitting this up on the type level is a good move). I suppose `assert RatingResult.has_answers(rating_result)` isn't too bad, just kind of annoying :/